### ISSUE TITLE
Bulk CSV upload: transaction, confirmation page

### ISSUE THREAD

**Issue by pbanaszkiewicz** [swcarpentry/amy#138]
We need to handle CSV uploads better.

Current workflow:

1. Prepare CSV file
2. Upload
3. Pray for it to work

We should have something less stressful instead:

1. Prepare CSV file
2. Upload
3. Make changes to fields with wrong values
4. Confirm
5. Save & be happy

Required changes:

* pack whole thing in database transaction
* make confirmation page
* make editable confirmation page

**Comment by wking** [comment]
On Wed, Jan 14, 2015 at 02:33:08AM -0800, Piotr Banaszkiewicz wrote:
> 1. Prepare CSV file
> 2. Upload

If we're preparing CSV by hand, we're doing it wrong (we should just
be using the Amy UI directly).  If CSV uploading doesn't work, we
should be fixing the CSV generator.  I doubt building a more
complicated CSV-upload-checker into amy is worth the trouble,
although…

> * pack whole thing in database transaction

… this makes sense to me (see my ATOMIC_REQUESTS suggestion [1]).

[1]: https://github.com/swcarpentry/amy/pull/106#issuecomment-68576325

**Comment by pbanaszkiewicz** [comment]
@wking the bulk-adding people feature has landed. It was really needed. I just made it more user-friendly and (I hope!) less error-prone.
I don't know if CSVs are being generated or prepared by humans. @gvwilson ?

**Comment by wking** [comment]
On Wed, Jan 14, 2015 at 01:16:29PM -0800, Piotr Banaszkiewicz wrote:
> the bulk-adding people feature has landed. It was really needed.

I'm still skeptical [1,2,3] ;).  But I'm not enough of a purist to
stand in the way of others who are willing to write, review, and
maintain things that I think could be done more easily ;).  I'll just
chime in with my point of view periodically to make sure folks aren't
confusing specific UI plans (like the bulk person upload) with the
usage requirements that those UIs are designed for (like adding people
to the database).

> I just made it more user-friendly and (I hope!) less error-prone.

Fair enough ;).  If you felt motivated to write the PR, I'm fine with
it landing.  Again, I just wanted to make sure we'd thought through
the “why do we need this, and could we accomplish that mission in a
simpler way” question.  If y'all feel satisfied that this is a
reasonably simple (or already working) solution to one of our usage
requirements, then I'm happy.

[1]: https://github.com/swcarpentry/amy/issues/27#issuecomment-67921209
[2]: https://github.com/swcarpentry/amy/issues/27#issuecomment-67968064
[3]: https://github.com/swcarpentry/amy/issues/27#issuecomment-67982553

**Comment by gvwilson** [comment]
Hi Piotr,

We get attendance data from many different sources in many different 
ways: scans of sign-in sheets, spreadsheets, CSV dumped directly from 
Eventbrite, and stuff typed in by hand.  CSV is the one format that 
everyone (technical or non-technical) seems to be able to work with, 
which is why we've gone with it.

I'll review this and give it a test drive tomorrow.

Thanks,
Greg

**Comment by pbanaszkiewicz** [comment]
@wking sorry if I sounded offending, I did not want to :) Honestly speaking, I'm feeling fine about #139 except for the latest commit (bb0a940167c842b9ecc8ad074335a42508451451) that introduces in-place manipulation of uploaded data. Everything else (data verification, confirmation page, transaction, etc.) seems to me like a decent code that can easily be read, understood and maintained.

If anyone has any idea how to improve bb0a940167c842b9ecc8ad074335a42508451451 (especially backend side, in `views.py`) I'd highly appreciate it or any other feedback.

@gvwilson alright, thanks!
